### ISSUE TITLE
Change deploy process

### ISSUE THREAD

**Issue by bmesuere** [dodona-edu/dodona#4673]
The goal of this plan is to limit the impact of changes by deploying as often as possible, preferably after each pull request merge. This would also allow further automation by using GitHub Actions. However, there are some drawbacks and unknowns that need to be addressed.

## Deployment process
Ideally, all deployments would be done through GitHub Actions. It should be possible to start a deployment manually and automatically when merging a PR. Ideally, a deployment would go through the following steps:
- Run all tests
- Test the deployment on naos (can we deploy the current production version first?)
- Deploy on dodona

## Steps to implement this

```[tasklist]
### Steps
- [ ]  add a new main branch
- [ ] replace all mentions of master by main
- [ ] set the "main branch" of this repo to main instead of develop
- [ ] change the branch protection rules on main/remove them on develop
- [ ] add a GitHub Action to deploy to production
- [ ] change the capistrano config: allow to deploy commit hashes to production, remove mentions of develop
- [ ] remove the hotfix creation script
```

```[tasklist]
### To discuss
- [ ] how will we implement rollback? Does capistrano offer something?
- [ ] can we automatically add some kind of version number to the footer to indicate the deployed version?
```

### What about dependabot PRs?
In this case, we probably only want to do a single deploy instead of one every merge. I see two possible strategies:
Only start a deployment if the PR has the bug, feature or enhancement label (and thus not on dependencies and chores).
Only deploy PRs that have a deploy label (adding the label could be automated)

### What about grouped PRs?
This is somewhat similar to the question above, but in this case I would use a different strategy and only deploy if the PR target is the main branch. Grouped PRs could then merge into an intermediate branch first.

### What about the exam period?
We tend not to do large releases during the exam period. Releasing more often would reduce the risk of a release, so not releasing during these periods would be less of an issue with this new strategy.

If we think that a PR might cause problems, we can still postpone it until the evening/weekend or after the exams. Note that the exam period becomes a very large part of the term if you combine secondary and higher education.

### Don’t we need version numbers for marketing?
We still want some sort of changelog. Creating a new GitHub release for each deployment will quickly become a mess. One solution might be to decouple GitHub releases from deployments. We could create a new release every month that contains all the changes from the previous month. We can then still document our changes with a topic in GitHub Discussions.

Release numbers could refer to the month, e.g. 2023.05. This release would be created at the beginning of May (or June). On the Dodona website, we probably want some indication of deployment, to know if a new version is still being deployed or not. The number there could be the deployment timestamp, or something of the form 2023.05.x.

### What about the develop branch?
Using this strategy would actually eliminate the need for the development branch (and hotfix branches). This would simplify our branches.

### What if something goes wrong?
Indeed, we would need a simple process for redeploying a previous working release. Creating a new PR with a revert commit is too cumbersome.

**Comment by bmesuere** [comment]
@niknetniko @chvp @jorg-vr @rien Can you comment on this before we start with the implementation?

**Comment by jorg-vr** [comment]
> how will we implement rollback? Does capistrano offer something?

I have actually recently learned it is quite easy to deploy a specific commit/branch/tag with capistrano
A rollback would thus simply be deploying an earlier tag

A revert of the pr would still also be advised, or a block on automated release in this, otherwise the code would still be included in the next release.

**Comment by jorg-vr** [comment]
I would not go fully automated directly

For example deploying on naos does not make much sense in a fully automated flow if we aren't going to use the check

How I would see it:
- trigger deploy action
- github action runs tests
- github action deploys on naos
- github action waits for confirmation
- Developer double checks impact on naos
- Developer confirms on github action
- Github action deploys

This trigger could also start on a merge. 
That would save us one manual action. But it would also create extra issues, such as multiple prs that are reviewed at once (which happens regularly with smaller issues) would each trigger a release flow. 
Also after a rollback to a previous release we should be cautious not to trigger a new automated release without the issue being fixed.

**Comment by jorg-vr** [comment]
We could achieve the same continuous deployment without the automatic trigger by simply making the decision to manually deploy after ever merge

**Comment by jorg-vr** [comment]
As discussed:

Phase 1:
- [ ] create new main branch
- [ ] Make main the github root
- [ ] Edit capistrano production deploy to use main
Manual deploy after each merge on main

**Comment by rien** [comment]
> > how will we implement rollback? Does capistrano offer something?
> 
> I have actually recently learned it is quite easy to deploy a specific commit/branch/tag with capistrano A rollback would thus simply be deploying an earlier tag

It is also possible to run [`cap production deploy:rollback`](https://capistranorb.com/documentation/getting-started/rollbacks/) which will set the linked folder (`current`) to the previous version. Since no pull or build steps are needed, this rollback will be faster.

However, this would mean that in cases where we rollback, the code that is on the main branch and on the server is not the same anymore.

**Comment by bmesuere** [comment]
@rien yes, you're right that with a rollback there would be a difference between the main branch and the deployment, but that shouldn't be an issue. I would use this in an emergency to fix a broken deploy. The next phase should of course be to actually fix the problem or do a revert. If the breaking deploy is not a big problem, we can skip the rollback and try to fix it immediately.

**Comment by bmesuere** [comment]
regarding showing a version in the footer: I think it would be easiest to add a step to capistrano to set an env variable with the date/time of the deployment (yy.MM.dd.hhmm ?). In rails, it's then easy to use that variable in the view.

**Comment by bmesuere** [comment]
Phase 2
- [ ] Enable branch protections on main
- [ ] Add a capistrano step to set an ENV variable with the deploy time
- [ ] Automatically run deploy script on PR merge (except if dependencies or no-deploy label is present)
